In [2]:
print("Transactions:", transactions.shape)
print("Users:", users.shape)
print("Merchants:", merchants.shape)
print("Fraud Labels:", fraud_labels.shape)
print("Data Dictionary:", data_dictionary.shape)

Transactions: (20000, 30)
Users: (2000, 13)
Merchants: (400, 9)
Fraud Labels: (20000, 11)
Data Dictionary: (63, 6)


In [3]:
transactions.head()


,transaction_id,user_id,receiver_id,receiver_type,amount,timestamp,date,hour_of_day,day_of_week,is_weekend,...,user_loyalty_score,new_device_flag,ip_location_mismatch,failed_attempts_last_24h,transaction_velocity,amount_deviation_score,is_fraud,recurring_payment_flag,balance_after_transaction,transaction_frequency_score
0,TXN0013171,USR01949,MRC0069,Merchant,1533.88,2024-01-01 00:21:17,2024-01-01,0,Monday,0,...,0.194,0,0,2,0.0,2.2467,0,1,36023.96,0.72
1,TXN0006480,USR01975,MRC0048,Merchant,1371.57,2024-01-01 00:36:29,2024-01-01,0,Monday,0,...,0.213,0,0,0,4.0,1.1773,0,0,65138.89,0.68
2,TXN0001430,USR01127,MRC0312,Merchant,500.50,2024-01-01 00:47:46,2024-01-01,0,Monday,0,...,0.381,0,0,0,1.0,0.1471,0,1,69347.11,0.86
3,TXN0003501,USR00827,USR01877,User,724.05,2024-01-01 00:52:07,2024-01-01,0,Monday,0,...,0.238,0,0,1,0.0,1.6019,0,0,52305.43,1.00
4,TXN0004423,USR00673,USR00534,User,128.22,2024-01-01 01:22:31,2024-01-01,1,Monday,0,...,0.187,0,0,5,1.0,0.3076,0,0,9601.13,0.48


In [4]:
transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 30 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   transaction_id               20000 non-null  object 
 1   user_id                      20000 non-null  object 
 2   receiver_id                  20000 non-null  object 
 3   receiver_type                20000 non-null  object 
 4   amount                       20000 non-null  float64
 5   timestamp                    20000 non-null  object 
 6   date                         20000 non-null  object 
 7   hour_of_day                  20000 non-null  int64  
 8   day_of_week                  20000 non-null  object 
 9   is_weekend                   20000 non-null  int64  
 10  is_night_transaction         20000 non-null  int64  
 11  time_since_last_txn_min      17665 non-null  float64
 12  transaction_type             20000 non-null  object 
 13  payment_app     

In [5]:
transactions.describe(include="all")

,transaction_id,user_id,receiver_id,receiver_type,amount,timestamp,date,hour_of_day,day_of_week,is_weekend,...,user_loyalty_score,new_device_flag,ip_location_mismatch,failed_attempts_last_24h,transaction_velocity,amount_deviation_score,is_fraud,recurring_payment_flag,balance_after_transaction,transaction_frequency_score
count,20000,20000,20000,20000,20000.000000,20000,20000,20000.000000,20000,20000.000000,...,20000.00000,20000.000000,20000.000000,20000.000000,19607.000000,19633.000000,20000.000000,20000.00000,20000.000000,20000.000000
unique,20000,2000,2227,2,NaN,19990,365,NaN,7,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
top,TXN0013171,USR01262,MRC0383,Merchant,NaN,2024-05-26 21:58:31,2024-03-02,NaN,Monday,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
freq,1,22,59,15119,NaN,2,77,NaN,3016,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
mean,NaN,NaN,NaN,NaN,876.846854,NaN,NaN,11.565050,NaN,0.281400,...,0.56031,0.064800,0.038450,0.537250,0.547101,0.653635,0.038150,0.30210,40413.740082,0.615067
std,NaN,NaN,NaN,NaN,1083.900556,NaN,NaN,6.867629,NaN,0.449693,...,0.25846,0.246179,0.192285,1.040705,0.999292,0.789019,0.191563,0.45918,22915.172130,0.246966
min,NaN,NaN,NaN,NaN,13.700000,NaN,NaN,0.000000,NaN,0.000000,...,0.10100,0.000000,0.000000,0.000000,0.000000,0.000100,0.000000,0.00000,500.750000,0.020000
25%,NaN,NaN,NaN,NaN,277.672500,NaN,NaN,6.000000,NaN,0.000000,...,0.33000,0.000000,0.000000,0.000000,0.000000,0.221500,0.000000,0.00000,20328.310000,0.420000
50%,NaN,NaN,NaN,NaN,539.955000,NaN,NaN,12.000000,NaN,0.000000,...,0.56900,0.000000,0.000000,0.000000,0.000000,0.444500,0.000000,0.00000,40398.730000,0.600000
75%,NaN,NaN,NaN,NaN,1061.510000,NaN,NaN,17.000000,NaN,1.000000,...,0.78000,0.000000,0.000000,1.000000,1.000000,0.727600,0.000000,1.00000,60474.840000,0.800000


In [6]:
datasets = {
    "transactions": transactions,
    "users": users,
    "merchants": merchants,
    "fraud_labels": fraud_labels
}

for name, df in datasets.items():
    print(f"\n{name.upper()}")
    print("Shape:", df.shape)
    print("Duplicate rows:", df.duplicated().sum())
    print("Missing values:")
    print(df.isnull().sum())


TRANSACTIONS
Shape: (20000, 30)
Duplicate rows: 0
Missing values:
transaction_id                    0
user_id                           0
receiver_id                       0
receiver_type                     0
amount                            0
timestamp                         0
date                              0
hour_of_day                       0
day_of_week                       0
is_weekend                        0
is_night_transaction              0
time_since_last_txn_min        2335
transaction_type                  0
payment_app                       0
device_type                       0
status                            0
user_city_tier                    0
user_kyc_status                   0
user_avg_monthly_txn              0
user_avg_txn_value                0
user_loyalty_score                0
new_device_flag                   0
ip_location_mismatch              0
failed_attempts_last_24h          0
transaction_velocity            393
amount_deviation_score          3

In [7]:
data_dictionary.head(20)

,file,column,description,dtype,example_values,notes
0,transactions.csv,transaction_id,Unique identifier for each transaction,string,TXN0000001,Primary key
1,transactions.csv,user_id,Unique sender/payer identifier,string,USR00042,Foreign key → users.csv
2,transactions.csv,receiver_id,Recipient — either another user or a merchant,string,MRC0023 / USR00088,Foreign key → merchants.csv or users.csv
3,transactions.csv,receiver_type,Whether the receiver is a User or Merchant,string,"User, Merchant",Categorical
4,transactions.csv,amount,Transaction amount in Indian Rupees (₹),float,"152.50, 4999.00","Log-normal distribution, max ₹1,00,000"
5,transactions.csv,timestamp,Full datetime of the transaction,datetime,2024-03-15 14:32:07,Format: YYYY-MM-DD HH:MM:SS
6,transactions.csv,date,Date portion of timestamp,date,2024-03-15,Format: YYYY-MM-DD
7,transactions.csv,hour_of_day,Hour when transaction occurred (0–23),int,0–23,Extracted from timestamp
8,transactions.csv,day_of_week,Day name of transaction,string,"Monday, Saturday",Extracted from timestamp
9,transactions.csv,is_weekend,Whether transaction occurred on a weekend,int (binary),"0, 1",1 = Saturday or Sunday


In [8]:
data_dictionary.columns

Index(['file', 'column', 'description', 'dtype', 'example_values', 'notes'], dtype='object')

In [9]:
data_dictionary.shape

(63, 6)

In [10]:
data_dictionary.to_string(index=False)

"            file                      column                                                              description        dtype                                       example_values                                                   notes\ntransactions.csv              transaction_id                                   Unique identifier for each transaction       string                                           TXN0000001                                             Primary key\ntransactions.csv                     user_id                                           Unique sender/payer identifier       string                                             USR00042                                 Foreign key → users.csv\ntransactions.csv                 receiver_id                            Recipient — either another user or a merchant       string                                   MRC0023 / USR00088                Foreign key → merchants.csv or users.csv\ntransactions.csv               rece

In [11]:
# Check the first 5 transactions
transactions.head()

,transaction_id,user_id,receiver_id,receiver_type,amount,timestamp,date,hour_of_day,day_of_week,is_weekend,...,user_loyalty_score,new_device_flag,ip_location_mismatch,failed_attempts_last_24h,transaction_velocity,amount_deviation_score,is_fraud,recurring_payment_flag,balance_after_transaction,transaction_frequency_score
0,TXN0013171,USR01949,MRC0069,Merchant,1533.88,2024-01-01 00:21:17,2024-01-01,0,Monday,0,...,0.194,0,0,2,0.0,2.2467,0,1,36023.96,0.72
1,TXN0006480,USR01975,MRC0048,Merchant,1371.57,2024-01-01 00:36:29,2024-01-01,0,Monday,0,...,0.213,0,0,0,4.0,1.1773,0,0,65138.89,0.68
2,TXN0001430,USR01127,MRC0312,Merchant,500.50,2024-01-01 00:47:46,2024-01-01,0,Monday,0,...,0.381,0,0,0,1.0,0.1471,0,1,69347.11,0.86
3,TXN0003501,USR00827,USR01877,User,724.05,2024-01-01 00:52:07,2024-01-01,0,Monday,0,...,0.238,0,0,1,0.0,1.6019,0,0,52305.43,1.00
4,TXN0004423,USR00673,USR00534,User,128.22,2024-01-01 01:22:31,2024-01-01,1,Monday,0,...,0.187,0,0,5,1.0,0.3076,0,0,9601.13,0.48


In [12]:
# Check all transaction column names
transactions.columns.tolist()

['transaction_id',
 'user_id',
 'receiver_id',
 'receiver_type',
 'amount',
 'timestamp',
 'date',
 'hour_of_day',
 'day_of_week',
 'is_weekend',
 'is_night_transaction',
 'time_since_last_txn_min',
 'transaction_type',
 'payment_app',
 'device_type',
 'status',
 'user_city_tier',
 'user_kyc_status',
 'user_avg_monthly_txn',
 'user_avg_txn_value',
 'user_loyalty_score',
 'new_device_flag',
 'ip_location_mismatch',
 'failed_attempts_last_24h',
 'transaction_velocity',
 'amount_deviation_score',
 'is_fraud',
 'recurring_payment_flag',
 'balance_after_transaction',
 'transaction_frequency_score']

In [13]:
# Check unique values in important categorical columns

print("Transaction Types:")
print(transactions["transaction_type"].value_counts())

print("\nPayment Apps:")
print(transactions["payment_app"].value_counts())

print("\nDevice Types:")
print(transactions["device_type"].value_counts())

print("\nTransaction Status:")
print(transactions["status"].value_counts())

print("\nReceiver Types:")
print(transactions["receiver_type"].value_counts())

Transaction Types:
transaction_type
P2M             7014
P2P             4881
Bill Payment    3033
Recharge        2063
EMI             1628
Subscription    1381
Name: count, dtype: int64

Payment Apps:
payment_app
GPay            6438
PhonePe         5864
Paytm           3738
BHIM            1595
Amazon Pay      1584
WhatsApp Pay     781
Name: count, dtype: int64

Device Types:
device_type
Android    13057
iOS         5405
Web         1538
Name: count, dtype: int64

Transaction Status:
status
Success    17628
Failed      1815
Pending      557
Name: count, dtype: int64

Receiver Types:
receiver_type
Merchant    15119
User         4881
Name: count, dtype: int64


In [14]:
# Check the fraud distribution

print(transactions["is_fraud"].value_counts())

print("\nFraud Percentage:")
print(transactions["is_fraud"].value_counts(normalize=True) * 100)

is_fraud
0    19237
1      763
Name: count, dtype: int64

Fraud Percentage:
is_fraud
0    96.185
1     3.815
Name: proportion, dtype: float64


In [15]:
fraud_summary = transactions.groupby("is_fraud").agg(
    transaction_count=("transaction_id", "count"),
    average_amount=("amount", "mean"),
    total_amount=("amount", "sum"),
    average_failed_attempts=("failed_attempts_last_24h", "mean"),
    average_velocity=("transaction_velocity", "mean"),
    average_deviation=("amount_deviation_score", "mean")
)
fraud_summary 

,transaction_count,average_amount,total_amount,average_failed_attempts,average_velocity,average_deviation
is_fraud,,,,,,
0,19237,874.416412,16821148.52,0.528461,0.527630,0.649034
1,763,938.123919,715788.55,0.758847,1.035952,0.769819


In [16]:
# Compare important fraud signals

fraud_signals = transactions.groupby("is_fraud")[
    [
        "new_device_flag",
        "ip_location_mismatch",
        "recurring_payment_flag",
        "is_night_transaction"
    ]
].mean() * 100

fraud_signals

,new_device_flag,ip_location_mismatch,recurring_payment_flag,is_night_transaction
is_fraud,,,,
0,5.905287,3.607631,30.243801,31.969642
1,20.969856,9.829620,29.357798,44.823067


In [17]:
fraud_by_type = transactions.groupby("transaction_type")["is_fraud"].mean() * 100

fraud_by_type.sort_values(ascending=False)

transaction_type
P2M             3.992016
Subscription    3.982621
Recharge        3.732429
P2P             3.728744
Bill Payment    3.692713
EMI             3.501229
Name: is_fraud, dtype: float64

In [18]:
fraud_by_app = transactions.groupby("payment_app")["is_fraud"].mean() * 100

fraud_by_app.sort_values(ascending=False)

payment_app
WhatsApp Pay    4.353393
Amazon Pay      4.166667
Paytm           3.986089
GPay            3.805530
PhonePe         3.751705
BHIM            3.072100
Name: is_fraud, dtype: float64

In [19]:
fraud_by_device = transactions.groupby("device_type")["is_fraud"].mean() * 100

fraud_by_device.sort_values(ascending=False)

device_type
Web        4.161248
Android    3.829364
iOS        3.681776
Name: is_fraud, dtype: float64

In [20]:
fraud_by_status = transactions.groupby("status")["is_fraud"].mean() * 100

fraud_by_status.sort_values(ascending=False)

status
Pending    5.026930
Success    3.795099
Failed     3.636364
Name: is_fraud, dtype: float64

🚨 New device: 20.97% fraud vs 5.91% normal
🚨 Night transaction: 44.82% fraud vs 31.97% normal
🚨 Higher transaction velocity: 1.04 vs 0.53
⚠️ IP mismatch: 9.83% vs 3.61%
⚠️ Pending status: 5.03% fraud rate
➖ Transaction type/device/app: relatively smaller differences

In [21]:
fraud_by_hour = transactions.groupby("hour_of_day")["is_fraud"].mean() * 100

fraud_by_hour

hour_of_day
0     4.967320
1     6.315789
2     5.162065
3     5.655836
4     5.012853
5     5.418719
6     2.205882
7     3.610108
8     2.004717
9     3.712871
10    2.637890
11    3.084223
12    4.255319
13    3.944316
14    2.368138
15    2.948113
16    2.243211
17    3.694875
18    4.327485
19    2.923264
20    2.680653
21    3.280680
22    4.310345
23    5.210918
Name: is_fraud, dtype: float64

In [22]:
fraud_by_hour.sort_values(ascending=False)

hour_of_day
1     6.315789
3     5.655836
5     5.418719
23    5.210918
2     5.162065
4     5.012853
0     4.967320
18    4.327485
22    4.310345
12    4.255319
13    3.944316
9     3.712871
17    3.694875
7     3.610108
21    3.280680
11    3.084223
15    2.948113
19    2.923264
20    2.680653
10    2.637890
14    2.368138
16    2.243211
6     2.205882
8     2.004717
Name: is_fraud, dtype: float64

Fraud rates tend to be higher during late-night/early-morning hours

In [25]:
fraud_by_day = transactions.groupby("day_of_week")["is_fraud"].mean() * 100

fraud_by_day.sort_values(ascending=False)

day_of_week
Saturday     3.997170
Friday       3.941056
Thursday     3.860029
Tuesday      3.855591
Wednesday    3.803768
Sunday       3.712960
Monday       3.547745
Name: is_fraud, dtype: float64

Fraud rates remain relatively consistent across the week, with Saturday showing the highest rate and Monday the lowest. However, the small difference suggests that day of week is not a strong standalone indicator of fraud

In [24]:
fraud_by_kyc = transactions.groupby("user_kyc_status")["is_fraud"].mean() * 100
fraud_by_kyc.sort_values(ascending=False)

user_kyc_status
Not Verified    5.691057
Verified        3.521453
Name: is_fraud, dtype: float64

Users with unverified KYC status show a significantly higher fraud rate (5.69%) compared with verified users (3.52%), suggesting that KYC verification may be an important risk indicator for fraud detection

In [27]:
fraud_by_city_tier = transactions.groupby("user_city_tier")["is_fraud"].mean() * 100
fraud_by_city_tier.sort_values(ascending=False)

user_city_tier
Tier 3    4.083748
Tier 1    3.882139
Tier 2    3.593263
Name: is_fraud, dtype: float64

Tier 3 users show the highest fraud rate (4.08%), while Tier 2 users have the lowest (3.59%). However, the relatively small difference suggests that city tier alone is not a strong predictor of fraudulent transactions

In [28]:
fraud_by_new_device = transactions.groupby("new_device_flag")["is_fraud"].mean() * 100
fraud_by_new_device.sort_values(ascending=False)

new_device_flag
1    12.345679
0     3.223909
Name: is_fraud, dtype: float64

Transactions made from new devices have a substantially higher fraud rate (12.35%) compared with transactions from existing devices (3.22%), making new-device activity a strong potential fraud-risk indicator

In [29]:
fraud_by_ip_mismatch = transactions.groupby("ip_location_mismatch")["is_fraud"].mean() * 100
fraud_by_ip_mismatch.sort_values(ascending=False)

ip_location_mismatch
1    9.752926
0    3.577557
Name: is_fraud, dtype: float64

Transactions with an IP location mismatch show a substantially higher fraud rate (9.75%) compared with transactions without a mismatch (3.58%), making IP location mismatch a strong potential fraud-risk indicator.
    

In [30]:
fraud_by_failed_attempts = transactions.groupby(
    "failed_attempts_last_24h"
)["is_fraud"].mean() * 100

fraud_by_failed_attempts.sort_values(ascending=False)

failed_attempts_last_24h
4    8.415842
5    6.511628
3    5.619413
2    4.574243
1    3.877052
0    3.458153
Name: is_fraud, dtype: float64

Fraud rates generally increase with the number of failed attempts in the previous 24 hours. Transactions associated with four failed attempts show the highest fraud rate at 8.42%, compared with 3.46% for transactions with no failed attempts

In [31]:
fraud_by_velocity = transactions.groupby(
    "transaction_velocity"
)["is_fraud"].mean() * 100

fraud_by_velocity.sort_values(ascending=False)

transaction_velocity
4.0    11.950791
3.0     6.666667
2.0     6.351183
1.0     5.036219
0.0     2.784847
Name: is_fraud, dtype: float64

Fraud rates increase substantially with transaction velocity. Transactions with a velocity score of 4 have the highest fraud rate at 11.95%, compared with 2.78% for velocity 0, making transaction velocity a strong potential fraud-risk indicator

In [34]:
import numpy as np

In [35]:
conditions = [
    transactions["amount_deviation_score"] < 0.5,
    transactions["amount_deviation_score"] < 1,
    transactions["amount_deviation_score"] < 2,
    transactions["amount_deviation_score"] < 3
]

labels = [
    "Low",
    "Moderate",
    "High",
    "Very High"
]

transactions["deviation_group"] = np.select(
    conditions,
    labels,
    default="Extreme"
)

fraud_by_deviation = (
    transactions.groupby("deviation_group", observed=True)["is_fraud"]
    .mean() * 100
)

fraud_by_deviation

deviation_group
Extreme      4.534314
High         5.050996
Low          3.340311
Moderate     3.949534
Very High    5.829596
Name: is_fraud, dtype: float64

Fraud rates generally increase as transaction amount deviation rises, reaching 5.83% for very-high deviation scores (2–3). However, the extreme deviation group shows a lower rate of 4.90%, indicating that the relationship is not strictly linear.

In [36]:
fraud_by_combined_risk = (
    transactions.groupby(
        ["new_device_flag", "ip_location_mismatch"]
    )["is_fraud"]
    .mean() * 100
)

fraud_by_combined_risk

new_device_flag  ip_location_mismatch
0                0                        2.976025
                 1                        9.353508
1                0                       12.200957
                 1                       16.666667
Name: is_fraud, dtype: float64

Transactions involving both a new device and an IP location mismatch showed the highest observed fraud rate (16.67%), compared with 2.98% for transactions with neither signal. This suggests that the combination of these two risk indicators may be particularly useful for identifying potentially fraudulent transactions

In [37]:
fraud_by_failed_attempts = (
    transactions.groupby("failed_attempts_last_24h")["is_fraud"]
    .mean() * 100
)

fraud_by_failed_attempts.sort_values(ascending=False)

failed_attempts_last_24h
4    8.415842
5    6.511628
3    5.619413
2    4.574243
1    3.877052
0    3.458153
Name: is_fraud, dtype: float64

Fraud rates generally increased with the number of failed attempts in the previous 24 hours. Transactions associated with four failed attempts had the highest observed fraud rate at 8.42%, compared with 3.46% for transactions with no failed attempts.

In [38]:
transactions["high_failed_attempts"] = (
    transactions["failed_attempts_last_24h"] >= 3
)

fraud_by_risk = (
    transactions.groupby(
        ["new_device_flag", "ip_location_mismatch", "high_failed_attempts"]
    )["is_fraud"]
    .mean() * 100
)

fraud_by_risk.sort_values(ascending=False)

new_device_flag  ip_location_mismatch  high_failed_attempts
1                1                     True                    50.000000
                 0                     True                    17.777778
                 1                     False                   15.000000
                 0                     False                   11.769759
0                1                     True                    10.714286
                                       False                    9.239940
                 0                     True                     5.502392
                                       False                    2.786581
Name: is_fraud, dtype: float64

The combination of multiple fraud-risk indicators produced substantially higher observed fraud rates. Transactions involving a new device, IP location mismatch, and three or more failed attempts had a 50% observed fraud rate, compared with 2.79% when none of these indicators were present. This highlights the potential value of combining multiple behavioral and security signals for fraud-risk identification.


In [39]:
risk_summary = (
    transactions.groupby(
        ["new_device_flag", "ip_location_mismatch", "high_failed_attempts"]
    )["is_fraud"]
    .agg(
        transaction_count="count",
        fraud_count="sum",
        fraud_rate="mean"
    )
)

risk_summary["fraud_rate"] = risk_summary["fraud_rate"] * 100

risk_summary.sort_values("fraud_rate", ascending=False)

transaction_count  \
new_device_flag ip_location_mismatch high_failed_attempts                      
1               1                    True                                  2   
                0                    True                                 90   
                1                    False                                40   
                0                    False                              1164   
0               1                    True                                 56   
                                     False                               671   
                0                    True                               1254   
                                     False                             16723   

                                                           fraud_count  \
new_device_flag ip_location_mismatch high_failed_attempts                
1               1                    True                            1   
                0                    True                           16   
                1                    False                           6   
                0                    False                         137   
0               1                    True                            6   
                                     False                          62   
                0                    True                           69   
                                     False                         466   

                                                           fraud_rate  
new_device_flag ip_location_mismatch high_failed_attempts              
1               1                    True                   50.000000  
                0                    True                   17.777778  
                1                    False                  15.000000  
                0                    False                  11.769759  
0               1                    True                   10.714286  
                                     False                   9.239940  
                0                    True                    5.502392  
                                     False                   2.786581

Combining multiple fraud-risk indicators reveals substantially higher fraud rates than transactions without these indicators. While the highest observed rate was 50%, it was based on only two transactions. More reliable groups also showed elevated fraud rates, particularly transactions involving a new device, where fraud rates exceeded 11%. This suggests that combining risk signals can provide stronger fraud-risk identification than relying on individual indicators alone.